src/main/java/com/minecolonies/coremod/commands/CommandEntryPointNew.java

```diff
@@ -0,0 +1,596 @@
+                                           @NotNull final String[] args,
+                                           @Nullable final BlockPos pos)
+    {
+        final ParsingResult parsingResult = getTabCompletionsAndParsingHolders(root, server, sender, args, pos);
+        return parsingResult.getTabCompletions();
+    }
+
+    @NotNull
+    private ParsingResult getTabCompletionsAndParsingHolders(
+                                           @NotNull final TreeNode<Menu> treeNode,
+                                           @NotNull final MinecraftServer server,
+                                           @NotNull final ICommandSender sender,
+                                           @NotNull final String[] args,
+                                           @Nullable final BlockPos pos)
+    {
+        if (treeNode.getData().getMenuType().isNavigationMenu())
+        {
+            final Map<String, TreeNode<Menu>> childs = getNavigationCommands(treeNode);
+            final String lowerCaseArg0 = args[0].toLowerCase();
+            if (args.length <= 1
+                  || !childs.containsKey(lowerCaseArg0))
+            {
+                final List<String> tabCompletions = childs.keySet().stream().filter(k -> k.startsWith(lowerCaseArg0)).collect(Collectors.toList());
+                if (childs.containsKey(lowerCaseArg0))
+                {
+                    final TreeNode<Menu> childTreeNode = childs.get(lowerCaseArg0);
+                    return new ParsingResult(tabCompletions, childTreeNode, (List<ActionArgument>) null, (String) null);
+                }
+                else
+                {
+                    return new ParsingResult(tabCompletions, treeNode, (List<ActionArgument>) null, lowerCaseArg0);
+                }
+            }
+            final TreeNode<Menu> child = childs.get(lowerCaseArg0);
+            if (null == child)
+            {
+                final List<String> tabCompletions = Collections.emptyList();
+                return new ParsingResult(tabCompletions, treeNode, (List<ActionArgument>) null, lowerCaseArg0);
+            }
+            final String[] newArgs = new String[args.length - 1];
+            System.arraycopy(args, 1, newArgs, 0, newArgs.length);
+            return getTabCompletionsAndParsingHolders(child, server, sender, newArgs, pos);
+        }
+        else
+        {
+            final List<ActionMenuHolder> parsedHolders = new ArrayList<>();
+            final Map<String, ActionMenuHolder> possibleActionCommands = getActionCommands(treeNode, parsedHolders);
+            @NotNull final List<ActionArgument> parsedActionArgumentList = new ArrayList<>();
+            return getTabCompletionsAndParsingHolders(parsedHolders, treeNode, parsedActionArgumentList, possibleActionCommands, server, sender, args, pos);
+        }
+    }
+
+    @NotNull
+    private ParsingResult getTabCompletionsAndParsingHolders(
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)